doc/source/Code-Style-Guide.rst

```diff
@@ -0,0 +1,746 @@
+
+In general, numerical data should be 32-bit (either float32 or int32 is
+acceptable), with the exception of absolute times, which cannot be
+handled with sufficient precision as 32-bit. Absolute times are
+therefore handled as 64-bit integers, in units of “seconds since
+1970-01-01 00:00:00”.
+
+Avoiding 64-bit floats
+^^^^^^^^^^^^^^^^^^^^^^
+
+The Python float and numpy default is 64 bit floating point, which
+equates to 15 or 16 significant digits. This is excessive for most of
+our problems, where e.g. temperature to the hundredth of a Kelvin or
+probabilities to 0.01 are good enough. Plugin code should avoid 64 bit
+quantities and arithmetic wherever possible and appropriate. 64 bit
+floating point is OK for example for Unix time values with non-integer
+seconds, but not for most physical quantities or probabilities.
+
+This means passing in ``dtype=np.float32`` to most numpy array
+constructor functions (e.g. ``array``, ``full``, ``zeros``, ``ones``,
+``arange``) and avoiding Python floating point numbers. You could use
+‘astype’ to cast if your array is already 64 bit, but it is best for
+performance to try to track down 64 bit computation at the places it
+comes in.
+
+.. code:: python
+
+
+   # Bad
+   foo = (bar + baz) / 2.0  # 2.0 is float64, so even if bar and baz are float32 foo will come out float64.
+   qux = np.ones((1, 2, 3))  # Default np array is float64.
+   wibble = np.array([wobble, wubble])  # Even if wobble and wubble are float32, no dtype is given, so float64
+   fred = float(garply)  # 'float' is 64 bit floating point
+
+   # Good
+   foo = (bar + baz) / np.float32(2.0)  # Or alternatively np.float32((bar + baz) / 2.0).
+   qux = np.ones((1, 2, 3), dtype=np.float32)
+   wibble = np.array([wobble, wubble], dtype=np.float32)
+   fred = np.float32(garply)
+
+Plugins
+-------
+
+Docstrings
+~~~~~~~~~~
+
+These should follow PEP8 and PEP257 - examples are shown in the example
+plugin code below.
+
+Note that private methods do not always required complete doc-strings,
+particularly if their behaviour is very obvious and the arguments to the
+method have been defined in other doc-strings. However, if the private
+method is complex, or arguments are being passed in that have not been
+defined elsewhere, a doc-string is advisable. If any argument needs
+defining for a private method, all arguments to that method should be
+defined in its docstring to avoid partial information.
+
+Due to the use of `Sphinx <http://www.sphinx-doc.org/en/stable/>`__ for
+documentation building, a number of `docstring section
+headers <http://sphinxcontrib-napoleon.readthedocs.io/en/latest/index.html#docstring-sections>`__
+are supported. Favoured docstring section headers are: \* Args:
+Compulsory arguments. \* Keyword Args : Keyword arguments. \* Raises:
+Exceptions raised. \* Returns: Variables returned by the function or
+method. \* References: Link to available documentation. \* Warns:
+Warnings raised.
```

Comment:
Another bulleted list that has been inlined.

Reply:
Fixed.